diff --git a/menu_token.module b/menu_token.module
--- a/menu_token.module
+++ b/menu_token.module
@@ -389,6 +389,9 @@
   if (!empty($config_menu)) {
     $configuration = unserialize($config_menu->configurationSerialized);
     $links[$relevant_link['id']][$replace_with] = $token_service->replace($links[$relevant_link['id']][$replace_with], [], ["configuration" => $configuration], $bubbleableMetadata);
+    if (is_null($links[$relevant_link['id']][$replace_with])) {
+      $links[$relevant_link['id']][$replace_with] = $token_service->replace($config_menu->linkid, [], ["configuration" => $configuration], $bubbleableMetadata);
+    }
     $links[$relevant_link['id']]["options"]["bubleble_metadata"] = $bubbleableMetadata;
   }
 }