diff --git a/draggableviews.module b/draggableviews.module
index 0950a5a..1a07574 100644
--- a/draggableviews.module
+++ b/draggableviews.module
@@ -110,8 +110,14 @@ function draggableviews_preprocess_views_view_table(&$variables) {
     $variables['rows'][$key]['columns'][$first_column]['content'][0]['field_output']['#markup'] = (string) (render($indent) . $columns_title);
   }
 
-  // Add table attributes.
-  $variables['attributes']['id'] = $draggableviews->getHtmlId();
+  // Get an array of field group titles.
+  $fieldGrouping = $draggableviews->fieldGrouping();
+  foreach ($fieldGrouping as $key => $row) {
+    if (trim(strip_tags($variables['title'])) == trim(strip_tags($row))) {
+      // Add table attributes.
+      $variables['attributes']['id'] = $draggableviews->getHtmlId($key);
+    }
+  }
 
   // Add rows attributes.
   foreach ($variables['rows'] as &$row) {
@@ -191,7 +197,9 @@ function draggableviews_views_submit(&$form, FormStateInterface $form_state) {
       if (isset($item['parent'])) {
         $record['parent'] = $item['parent'];
       }
-      $connection->insert('draggableviews_structure')->fields($record)->execute();
+      $connection->insert('draggableviews_structure')
+        ->fields($record)
+        ->execute();
       $weight++;
     }
     // We invalidate the entity list cache, so other views are also aware of the
@@ -202,12 +210,15 @@ function draggableviews_views_submit(&$form, FormStateInterface $form_state) {
       return $info['relationship_id'] == 'none';
     }));
     $entity_type_id = reset($result);
-    $list_cache_tags = \Drupal::entityTypeManager()->getDefinition($entity_type_id)->getListCacheTags();
+    $list_cache_tags = \Drupal::entityTypeManager()
+      ->getDefinition($entity_type_id)
+      ->getListCacheTags();
     Cache::invalidateTags($list_cache_tags);
   }
   catch (\Exception $e) {
     $transaction->rollback();
-    \Drupal::logger('draggableviews')->error('Failed with @message', ['@message' => $e->getMessage()]);
+    \Drupal::logger('draggableviews')
+      ->error('Failed with @message', ['@message' => $e->getMessage()]);
     drupal_set_message(t('There was an error while saving the data. Please, try gain.'), 'warning');
   }
 }
diff --git a/src/DraggableViews.php b/src/DraggableViews.php
index 7cfd4f7..f699b16 100644
--- a/src/DraggableViews.php
+++ b/src/DraggableViews.php
@@ -73,11 +73,23 @@ class DraggableViews {
     return $this->view->result[$index]->$name;
   }
 
+  /**
+   * Return array of field groups titles.
+   */
+  public function fieldGrouping() {
+    $fieldGrouping = [];
+    $sets = $this->view->style_plugin->renderGrouping($this->view->result, $this->view->style_plugin->options['grouping'], FALSE);
+    foreach ($sets as $title => $rows) {
+      $fieldGrouping[] = $title;
+    }
+    return $fieldGrouping;
+  }
+
   /**
    * Get HTML id for draggableviews table.
    */
-  public function getHtmlId() {
-    return Html::getId('draggableviews-table-' . $this->view->id() . '-' . $this->view->current_display);
+  public function getHtmlId($index) {
+    return Html::getId('draggableviews-table-' . $this->view->id() . '-' . $this->view->current_display . '-' . $index);
   }
 
 }
diff --git a/src/Plugin/views/field/DraggableViewsField.php b/src/Plugin/views/field/DraggableViewsField.php
index 59a04dc..ad65152 100755
--- a/src/Plugin/views/field/DraggableViewsField.php
+++ b/src/Plugin/views/field/DraggableViewsField.php
@@ -38,6 +38,7 @@ class DraggableViewsField extends BulkForm {
    * @var \Drupal\Core\Language\LanguageManagerInterface
    */
   protected $languageManager;
+
   /**
    * The Current user.
    *
@@ -146,15 +147,19 @@ class DraggableViewsField extends BulkForm {
     }
 
     if ($this->currentUser->hasPermission('access draggableviews')) {
-      $options = [
-        'table_id' => $draggableviews->getHtmlId(),
-        'action' => 'match',
-        'relationship' => 'group',
-        'group' => 'draggableviews-parent',
-        'subgroup' => 'draggableviews-parent',
-        'source' => 'draggableviews-id',
-      ];
-      drupal_attach_tabledrag($form, $options);
+      // Get an array of field group titles.
+      $fieldGrouping = $draggableviews->fieldGrouping();
+      foreach ($fieldGrouping as $key => $row) {
+        $options = [
+          'table_id' => $draggableviews->getHtmlId($key),
+          'action' => 'match',
+          'relationship' => 'parent',
+          'group' => 'draggableviews-parent',
+          'subgroup' => 'draggableviews-parent',
+          'source' => 'draggableviews-id',
+        ];
+        drupal_attach_tabledrag($form, $options);
+      }
     }
   }
 
